File: copilot-connector/sample_codes/common-patterns/acl-configuration.cs
// ACL configuration patterns: user, group, everyone, external groups, deny
using Microsoft.Graph;
using Microsoft.Graph.Models.ExternalConnectors;

public class AclConfiguration
{
    private readonly GraphServiceClient _graphClient;
    private readonly string _connectionId;

    public AclConfiguration(GraphServiceClient graphClient, string connectionId)
    {
        _graphClient = graphClient;
        _connectionId = connectionId;
    }

    /// <summary>
    /// Public item visible to all tenant users.
    /// </summary>
    public List<AclEntry> PublicAcl() => new()
    {
        new AclEntry
        {
            Type = AclType.Everyone,
            Value = "everyone",
            AccessType = AccessType.Grant
        }
    };

    /// <summary>
    /// Public item visible to all tenant users except guests.
    /// </summary>
    public List<AclEntry> InternalOnlyAcl() => new()
    {
        new AclEntry
        {
            Type = AclType.EveryoneExceptGuests,
            Value = "everyoneExceptGuests",
            AccessType = AccessType.Grant
        }
    };

    /// <summary>
    /// Item restricted to specific Entra ID users.
    /// Values MUST be Entra object IDs (GUIDs), not emails or UPNs.
    /// </summary>
    public List<AclEntry> UserRestrictedAcl(params string[] userEntraIds) =>
        userEntraIds.Select(id => new AclEntry
        {
            Type = AclType.User,
            Value = id,
            AccessType = AccessType.Grant
        }).ToList();

    /// <summary>
    /// Item restricted to members of specific Entra ID groups.
    /// Prefer group-based ACLs over user-based for maintainability.
    /// </summary>
    public List<AclEntry> GroupRestrictedAcl(params string[] groupEntraIds) =>
        groupEntraIds.Select(id => new AclEntry
        {
            Type = AclType.Group,
            Value = id,
            AccessType = AccessType.Grant
        }).ToList();

    /// <summary>
    /// Combination: grant to all, deny to specific users.
    /// Deny ALWAYS overrides grant.
    /// </summary>
    public List<AclEntry> EveryoneExceptAcl(params string[] deniedUserIds)
    {
        var acl = new List<AclEntry>
        {
            new AclEntry
            {
                Type = AclType.Everyone,
                Value = "everyone",
                AccessType = AccessType.Grant
            }
        };

        acl.AddRange(deniedUserIds.Select(id => new AclEntry
        {
            Type = AclType.User,
            Value = id,
            AccessType = AccessType.Deny
        }));

        return acl;
    }

    /// <summary>
    /// Item restricted to an external group (for non-Entra ID permissions).
    /// External groups must be created first via the group sync API.
    /// </summary>
    public List<AclEntry> ExternalGroupAcl(string externalGroupId) => new()
    {
        new AclEntry
        {
            Type = AclType.ExternalGroup,
            Value = externalGroupId,
            AccessType = AccessType.Grant
        }
    };

    // --- External Group Management ---

    /// <summary>
    /// Create an external group for non-Entra ID permission structures.
    /// Use for Salesforce roles, ServiceNow groups, custom RBAC, etc.
    /// </summary>
    public async Task<ExternalGroup> CreateExternalGroupAsync(
        string groupId, string displayName, string? description = null)
    {
        var group = new ExternalGroup
        {
            Id = groupId,
            DisplayName = displayName,
            Description = description
        };

        return await _graphClient.External.Connections[_connectionId]
            .Groups
            .PostAsync(group) ?? throw new Exception("Failed to create external group");
    }

    /// <summary>
    /// Add an Entra ID user to an external group.
    /// </summary>
    public async Task AddUserToExternalGroupAsync(
        string groupId, string userEntraId)
    {
        var member = new Identity
        {
            OdataType = "#microsoft.graph.externalConnectors.identity",
            Id = userEntraId,
            Type = IdentityType.User
        };

        await _graphClient.External.Connections[_connectionId]
            .Groups[groupId]
            .Members
            .PostAsync(member);
    }

    /// <summary>
    /// Add an Entra ID group to an external group (nested groups).
    /// </summary>
    public async Task AddGroupToExternalGroupAsync(
        string externalGroupId, string entraGroupId)
    {
        var member = new Identity
        {
            OdataType = "#microsoft.graph.externalConnectors.identity",
            Id = entraGroupId,
            Type = IdentityType.Group
        };

        await _graphClient.External.Connections[_connectionId]
            .Groups[externalGroupId]
            .Members
            .PostAsync(member);
    }
}

File: copilot-connector/sample_codes/typescript/common-patterns/acl-configuration.ts
// ACL configuration patterns: user, group, everyone, external groups, deny
//
// Prerequisites: npm install @microsoft/microsoft-graph-client @azure/identity

import { Client } from "@microsoft/microsoft-graph-client";

/** ACL entry controlling who can see an item. */
interface AclEntry {
  type: "everyone" | "everyoneExceptGuests" | "user" | "group" | "externalGroup";
  value: string;
  accessType: "grant" | "deny";
}

/** An external group definition. */
interface ExternalGroup {
  id: string;
  displayName: string;
  description?: string;
}

/**
 * Provides ACL configuration helpers and external group management
 * for Microsoft Graph external connections.
 */
class AclConfiguration {
  constructor(
    private readonly graphClient: Client,
    private readonly connectionId: string
  ) {}

  // --- ACL Builders ---

  /** Public item visible to all tenant users. */
  publicAcl(): AclEntry[] {
    return [
      {
        type: "everyone",
        value: "everyone",
        accessType: "grant",
      },
    ];
  }

  /** Public item visible to all tenant users except guests. */
  internalOnlyAcl(): AclEntry[] {
    return [
      {
        type: "everyoneExceptGuests",
        value: "everyoneExceptGuests",
        accessType: "grant",
      },
    ];
  }

  /**
   * Item restricted to specific Entra ID users.
   * Values MUST be Entra object IDs (GUIDs), not emails or UPNs.
   */
  userRestrictedAcl(...userEntraIds: string[]): AclEntry[] {
    return userEntraIds.map((id) => ({
      type: "user" as const,
      value: id,
      accessType: "grant" as const,
    }));
  }

  /**
   * Item restricted to members of specific Entra ID groups.
   * Prefer group-based ACLs over user-based for maintainability.
   */
  groupRestrictedAcl(...groupEntraIds: string[]): AclEntry[] {
    return groupEntraIds.map((id) => ({
      type: "group" as const,
      value: id,
      accessType: "grant" as const,
    }));
  }

  /**
   * Combination: grant to all, deny to specific users.
   * Deny ALWAYS overrides grant.
   */
  everyoneExceptAcl(...deniedUserIds: string[]): AclEntry[] {
    const acl: AclEntry[] = [
      {
        type: "everyone",
        value: "everyone",
        accessType: "grant",
      },
    ];

    for (const id of deniedUserIds) {
      acl.push({
        type: "user",
        value: id,
        accessType: "deny",
      });
    }

    return acl;
  }

  /**
   * Item restricted to an external group (for non-Entra ID permissions).
   * External groups must be created first via the group sync API.
   */
  externalGroupAcl(externalGroupId: string): AclEntry[] {
    return [
      {
        type: "externalGroup",
        value: externalGroupId,
        accessType: "grant",
      },
    ];
  }

  // --- External Group Management ---

  /**
   * Create an external group for non-Entra ID permission structures.
   * Use for Salesforce roles, ServiceNow groups, custom RBAC, etc.
   */
  async createExternalGroup(
    groupId: string,
    displayName: string,
    description?: string
  ): Promise<ExternalGroup> {
    const group = await this.graphClient
      .api(`/external/connections/${this.connectionId}/groups`)
      .post({
        id: groupId,
        displayName,
        description,
      });

    return group as ExternalGroup;
  }

  /**
   * Add an Entra ID user to an external group.
   */
  async addUserToExternalGroup(
    groupId: string,
    userEntraId: string
  ): Promise<void> {
    await this.graphClient
      .api(
        `/external/connections/${this.connectionId}/groups/${groupId}/members`
      )
      .post({
        "@odata.type": "#microsoft.graph.externalConnectors.identity",
        id: userEntraId,
        type: "user",
      });
  }

  /**
   * Add an Entra ID group to an external group (nested groups).
   */
  async addGroupToExternalGroup(
    externalGroupId: string,
    entraGroupId: string
  ): Promise<void> {
    await this.graphClient
      .api(
        `/external/connections/${this.connectionId}/groups/${externalGroupId}/members`
      )
      .post({
        "@odata.type": "#microsoft.graph.externalConnectors.identity",
        id: entraGroupId,
        type: "group",
      });
  }
}

export { AclConfiguration, AclEntry, ExternalGroup };
